util/src/trie/memorytriedb.rs

```diff
@@ -0,0 +1,681 @@
+use super::node::Node as RlpNode;
+
+use ::{Bytes, FixedHash, HashDB, H256, SHA3_NULL_RLP};
+use ::nibbleslice::NibbleSlice;
+use ::rlp::{Rlp, View};
+
+use std::collections::VecDeque;
+use std::mem;
+use std::ops::{Index, IndexMut};
+
+// For lookups into the Node storage buffer.
+// This is deliberately non-copyable.
+struct StorageHandle(usize);
+
+// Handles to nodes in the trie.
+enum NodeHandle {
+	/// Loaded into memory.
+	InMemory(StorageHandle),
+	/// Either a hash or an inline node
+	Hash(H256),
+}
+
+impl From<StorageHandle> for NodeHandle {
+	fn from(handle: StorageHandle) -> Self {
+		NodeHandle::InMemory(handle)
+	}
+}
+
+impl From<H256> for NodeHandle {
+	fn from(hash: H256) -> Self {
+		NodeHandle::Hash(hash)
+	}
+}
+
+fn empty_children() -> [Option<NodeHandle>; 16] {
+	[
+		None, None, None, None, None, None, None, None,
+		None, None, None, None, None, None, None, None,
+	]
+}
+
+/// Node types in the Trie.
+enum Node {
+	/// Empty node.
+	Empty,
+	/// A leaf node contains the end of a key and a value.
+	/// This key is encoded from a `NibbleSlice`, meaning it contains
+	/// a flag indicating it is a leaf.
+	Leaf(Bytes, Bytes),
+	/// An extension contains a shared portion of a key and a child node.
+	/// The shared portion is encoded from a `NibbleSlice` meaning it contains
+	/// a flag indicating it is an extension.
+	/// The child node is always a branch.
+	Extension(Bytes, NodeHandle),
+	/// A branch has up to 16 children and an optional value.
+	Branch([Option<NodeHandle>; 16], Option<Bytes>)
+}
+
+impl Node {
+	// load an inline node into memory or get the hash to do the lookup later.
+	fn inline_or_hash(node: &[u8], db: &HashDB, storage: &mut NodeStorage) -> NodeHandle {
+		let r = Rlp::new(node);
+		if r.is_data() && r.size() < 32 {
+			let child = Node::from_rlp(r.data(), db, storage);
+			NodeHandle::InMemory(storage.alloc(child))
+		} else {
+			NodeHandle::Hash(r.as_val::<H256>())
+		}
+	}
+
+	// decode a node from rlp without getting its children.
+	fn from_rlp(rlp: &[u8], db: &HashDB, storage: &mut NodeStorage) -> Self {
+		match RlpNode::decoded(rlp) {
+			RlpNode::Empty => Node::Empty,
+			RlpNode::Leaf(k, v) => Node::Leaf(k.encoded(true), v.to_owned()),
+			RlpNode::Extension(k, v) => {
+				let key = k.encoded(false);
+
+				Node::Extension(key, Self::inline_or_hash(v, db, storage))
+			}
+			RlpNode::Branch(children_rlp, v) => {
+				let val = v.map(|x| x.to_owned());
+				let mut children = empty_children();
+
+				for i in 0..16 {
+					let raw = children_rlp[i];
+					let child_rlp = Rlp::new(raw);
+					if !child_rlp.is_empty()  {
+						children[i] = Some(Self::inline_or_hash(raw, db, storage));
+					}
+				}
+
+				Node::Branch(children, val)
+			}
+		}
+	}
+
+	// encode a node to RLP. if it has child nodes, remove them from the storage,
+	// encode them, and write them to the DB.
+	//
+	// TODO: parallelize
+	fn to_rlp(self, db: &mut HashDB, storage: &mut NodeStorage) -> Bytes {
```

Comment:
returning `ElasticArray1024` here would save lots of memory allocations
`stream.as_raw()`